packages/@aws-cdk/aws-cognito/lib/user-pool.ts

```diff
@@ -0,0 +1,412 @@
+   */
+  Locale = 'locale',
+
+  /**
+   * Middle name(s) of the End-User.
+   * Note that in some cultures, people can have multiple middle names;
+   * all can be present, with the names being separated by space characters.
+   * Also note that in some cultures, middle names are not used.
+   */
+  MiddleName = 'middle_name',
+
+  /**
+   * End-User's full name in displayable form including all name parts,
+   * possibly including titles and suffixes, ordered according to the End-User's locale and preferences.
+   */
+  Name = 'name',
+
+  /**
+   * Casual name of the End-User that may or may not be the same as the given_name.
+   * For instance, a nickname value of Mike might be returned alongside a given_name value of Michael.
+   */
+  Nickname = 'nickname',
+
+  /**
+   * End-User's preferred telephone number.
+   * E.164 [E.164] is RECOMMENDED as the format of this Claim, for example, +1 (425) 555-1212 or +56 (2) 687 2400.
+   * If the phone number contains an extension, it is RECOMMENDED that the extension be represented using the
+   * RFC 3966 [RFC3966] extension syntax, for example, +1 (604) 555-1234;ext=5678.
+   */
+  PhoneNumber = 'phone_number',
+
+  /**
+   * URL of the End-User's profile picture.
+   * This URL MUST refer to an image file (for example, a PNG, JPEG, or GIF image file),
+   * rather than to a Web page containing an image.
+   * Note that this URL SHOULD specifically reference a profile photo of the End-User
+   * suitable for displaying when describing the End-User, rather than an arbitrary photo taken by the End-User
+   */
+  Picture = 'picture',
+
+  /**
+   * Shorthand name by which the End-User wishes to be referred to.
+   */
+  PreferredUsername = 'preferred_username',
+
+  /**
+   * URL of the End-User's profile page. The contents of this Web page SHOULD be about the End-User.
+   */
+  Profile = 'profile',
+
+  /**
+   * The End-User's time zone
+   */
+  Timezone = 'timezone',
+
+  /**
+   * Time the End-User's information was last updated.
+   * Its value is a JSON number representing the number of seconds from 1970-01-01T0:0:0Z
+   * as measured in UTC until the date/time.
+   */
+  UpdatedAt = 'updated_at',
+
+  /**
+   * URL of the End-User's Web page or blog.
+   * This Web page SHOULD contain information published by the End-User or an organization that the End-User is affiliated with.
+   */
+  Website = 'website'
+}
+
+/**
+ * Methods of user sign-in
+ */
+export enum SignInType {
+  /**
+   * End-user will sign in with a username, with optional aliases
+   */
+  Username,
+
+  /**
+   * End-user will sign in using an email address
+   */
+  Email,
+
+  /**
+   * End-user will sign in using a phone number
+   */
+  Phone,
+
+  /**
+   * End-user will sign in using either an email address or phone number
+   */
+  EmailOrPhone
+}
+
+export interface UserPoolTriggers {
+  /**
+   * Creates an authentication challenge.
+   */
+  createAuthChallenge?: lambda.IFunction;
```

Comment:
this looks to me like something we can also add as `On` methods in the user pool to support a more event-based API:

```ts
userPool.onCreateAuthChallenge(lambda);
```

Reply:
I like that idea!

@RomainMuller if implemented, would it be fine to also attach the service grants in these calls? It may make the side-effect easier to document than if it's just automatically generated/attached in the user pool constructor.

Reply:
@eladb should `on` methods be in addition to props, or should they replace props?

Reply:
@eladb should the `on` method replace the ability to pass the lambdas in `props`, or be available in addition to?

Reply:
In addition of course!